prophyc/generators/base.py

```diff
@@ -0,0 +1,87 @@
+import os
+from prophyc import model
+
+
+class GenerateError(Exception):
+    pass
+
+
+class GeneratorBase(object):
```

Comment:
It's not clear what part of this base class is subject to be:
1. overriden by derived class
2. used by derived class
3. used exclusively by base class

Maybe splitting this class to "abstract base class" which has only methods subject to overriding and "base class" which implements behavior using the "abstract" class would clear the situation up. `make_path` and `write_file` don't depend on being members of `GeneratorBase` - I'd move them outside this class entirely as free private functions to make it clear that GeneratorBase's only interface towards user is `__init__` and `serialize`.

If we have 4 entities:
`GeneratorAbc <- Generator <- PythonGenerator`
`GeneratorUser`
Then:
1. PythonGenerator implements GeneratorAbc (`file_translators` + `check_nodes`)
2. GeneratorUser uses Generator (`__init__` + `serialize`)

```
class GeneratorAbc(object):
    '''A bit of prose explaining that it defines requirements for derived Generator'''
    file_translators = {}

    def check_nodes(self, nodes):
        '''A bit of prose'''
        pass

class GeneratorBase(GeneratorAbc):
    def __init__(self, output_dir="."):
        self.output_dir = output_dir

    def serialize(self, nodes, base_name):
        self.check_nodes(nodes)

        for extension, translator_class in self.file_translators.items():
            translator = translator_class()
            contents = translator(nodes, base_name)
            file_path = self.make_path(base_name, extension)
            self.write_file(file_path, contents)
```